src/ht.c

```diff
@@ -1,239 +1,184 @@
-		hash_address = hash % ht->size;
-		do {
-			SdbHashEntry *entry = ht->table + hash_address;
-			if (entry_is_free (entry))
-				return NULL;
-			if (entry_is_present (entry) && entry->hash == hash)
-				return entry;
-			double_hash = hash % ht->rehash;
-			if (double_hash == 0)
-				double_hash = 1;
-			hash_address = (hash_address + double_hash) % ht->size;
-		} while (hash_address != hash % ht->size);
+SdbHash* internal_ht_new(ut64 size, HashFunction hashfunction,
+		ListComparator comparator, DupKey keydup, DupValue valdup,
+		HtKvFreeFunc pair_free) {
+	SdbHash* ht;
+	ut32 i;
+	ht = calloc (1, sizeof(*ht));
+	if (ht != NULL) {
+		ht->size = size;
+		ht->count = 0;
+		ht->load_factor = 1;
+		ht->hashfn = hashfunction;
+		ht->cmp = comparator;
+		ht->dupkey = keydup;
+		ht->dupvalue = valdup;
+		ht->table = calloc (ht->size, sizeof (SdbList*));
+		if (ht->table != NULL) {
+			for (i = 0; i < ht->size; ++i) {
+				ht->table[i] = ls_new ();
+				if (pair_free) {
+					ht->table[i]->free = (SdbListFree)pair_free;
+				}
+			}
+		} else {
+			free (ht);
+			return NULL;
+		}
 	}
-	return NULL;
+	return ht;
 }
 
-static int rehash = 0;
-static void ht_rehash(SdbHash *ht, ut32 new_size_index) {
-	SdbHash old_ht = *ht;
-	SdbHashEntry *e;
-	if (!ht || new_size_index >= ARRAY_SIZE (hash_sizes))
-		return;
-	// XXX: This code is redupped! fuck't
-	ht->table = calloc (hash_sizes[new_size_index].size, sizeof (*ht->table));
-	if (!ht->table)
-		return;
-	rehash = 1;
-	ht->size_index = new_size_index;
-	ht->size = hash_sizes[ht->size_index].size;
-	ht->rehash = hash_sizes[ht->size_index].rehash;
-	ht->max_entries = hash_sizes[ht->size_index].max_entries;
-	ht->entries = 0;
-	ht->deleted_entries = 0;
-	for (e = old_ht.table; e != old_ht.table + old_ht.size; e++) {
-		if (entry_is_present (e))
-			ht_insert (ht, e->hash, e->data, e->iter);
-	}
-	free (old_ht.table);
-rehash = 0;
+SdbHash* ht_new(HashFunction hashfunction, ListComparator comparator,
+		DupKey keydup, DupValue valdup, HtKvFreeFunc pair_free) {
+	ut64 initial_size = 1024;
+	return internal_ht_new (initial_size, hashfunction, comparator, keydup,
+			valdup, pair_free);
 }
 
-SdbHash* ht_new(SdbListFree f) {
-	SdbHash *ht = R_NEW (SdbHash);
-	if (!ht) return NULL;
-	// TODO: use slices here
-	ht->list = ls_new ();
-	ht->list->free = f;
-	ht->size = hash_sizes[0].size;
-	ht->table = calloc (ht->size, sizeof (*ht->table));
-	if (!ht->table) {
-		free (ht);
-		return NULL;
+void ht_destroy(SdbHash* ht) {
+	ut32 i;
+	for (i = 0; i < ht->size; ++i) {
+		ls_free (ht->table[i]);
 	}
-	ht->size_index = 0;
-	ht->entries = 0;
-	ht->deleted_entries = 0;
-	ht->rehash = hash_sizes[ht->size_index].rehash;
-	ht->max_entries = hash_sizes[ht->size_index].max_entries;
-	return ht;
+	free (ht->table);
+	free (ht);
 }
 
-void ht_free(SdbHash *ht) {
-	if (ht) {
-		free (ht->table);
-		ls_free (ht->list);
-		free (ht);
+// Increases the size of the hashtable by 2.
+static void internal_ht_grow(SdbHash* ht) {
```

Comment:
when we hit num entries == num buckets, this is the cannonical way to do hashtable growth.